File: scenarios/budget-limits/python/agent.py
#!/usr/bin/env python3
"""
Budget-limits scenario — Agent Assembly examples

Demonstrates how Agent Assembly enforces budget guardrails on governed tool calls.
No API keys or external services are required to run this example.

Usage:
    python agent.py
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# Per-call costs matching policy.yaml
TOOL_COSTS: dict[str, float] = {
    "web_search": 0.05,
    "query_database": 0.10,
    "call_external_api": 0.20,
    "generate_report": 0.25,
}

# Session budget ceiling matching policy.yaml budget.max_cost
BUDGET_LIMIT: float = 0.50


# ---------------------------------------------------------------------------
# Minimal Agent Assembly SDK stubs used in this offline example.
# In a real integration replace these with:
#   from agent_assembly import AssemblyClient, BudgetPolicy
# ---------------------------------------------------------------------------


class BudgetExceededError(PermissionError):
    pass


@dataclass
class BudgetTracker:
    max_cost: float
    _spent: float = field(default=0.0, init=False)

    @property
    def spent(self) -> float:
        return self._spent

    @property
    def remaining(self) -> float:
        return max(0.0, self.max_cost - self._spent)

    def charge(self, cost: float) -> None:
        self._spent += cost

    def status(self) -> str:
        pct = (self._spent / self.max_cost) * 100
        return f"spent=${self._spent:.2f} / limit=${self.max_cost:.2f} ({pct:.0f}%)"


class AssemblyClient:
    """Simulates the Agent Assembly governed runtime with budget enforcement."""

    def __init__(self, agent_id: str, budget: BudgetTracker) -> None:
        self.agent_id = agent_id
        self._budget = budget

    def call_tool(self, tool: str, **inputs: Any) -> dict[str, Any]:
        cost = TOOL_COSTS.get(tool, 0.01)

        if cost > self._budget.remaining:
            raise BudgetExceededError(
                f"Budget exceeded: tool '{tool}' costs ${cost:.2f} but "
                f"only ${self._budget.remaining:.2f} remains "
                f"(spent=${self._budget.spent:.2f} / limit=${self._budget.max_cost:.2f})"
            )

        self._budget.charge(cost)
        print(
            f"  [BUDGET] charged ${cost:.2f} for '{tool}'"
            f"  →  {self._budget.status()}"
        )
        return {"status": "ok", "tool": tool, "data": f"<result of {tool}>"}


# ---------------------------------------------------------------------------
# Example agent
# ---------------------------------------------------------------------------


def run() -> None:
    print("=== Agent Assembly — Budget Limits Example ===\n")
    print(f"Policy: max_cost=${BUDGET_LIMIT:.2f} per session (see policy.yaml)\n")

    budget = BudgetTracker(max_cost=BUDGET_LIMIT)
    client = AssemblyClient(agent_id="example-agent-001", budget=budget)

    calls: list[tuple[str, dict[str, Any]]] = [
        ("web_search", {"query": "latest AI news"}),          # $0.05 → $0.05
        ("query_database", {"table": "customers"}),           # $0.10 → $0.15
        ("call_external_api", {"endpoint": "/v1/report"}),    # $0.20 → $0.35
        ("web_search", {"query": "weather forecast"}),        # $0.05 → $0.40
        ("generate_report", {"format": "pdf"}),               # $0.25 → exceeds $0.50
        ("call_external_api", {"endpoint": "/v2/sync"}),      # $0.20 → blocked — not enough left
    ]

    print("--- Running tool calls against budget ---\n")

    for tool, kwargs in calls:
        cost = TOOL_COSTS.get(tool, 0.01)
        args_str = ", ".join(f"{k}={v!r}" for k, v in kwargs.items())
        print(f"  → {tool}({args_str})  [cost=${cost:.2f}]")
        try:
            result = client.call_tool(tool, **kwargs)
            print(f"    ✓ allowed  →  {result}\n")
        except BudgetExceededError as exc:
            print(f"    ✗ denied   →  {exc}\n")

    print(f"\nFinal budget state: {budget.status()}")


if __name__ == "__main__":
    run()

File: scenarios/budget-limits/node/agent.js
#!/usr/bin/env node
/**
 * Budget-limits scenario — Agent Assembly examples
 *
 * Demonstrates how Agent Assembly enforces budget guardrails on governed tool calls.
 * No API keys or external services are required to run this example.
 *
 * Usage:
 *   node agent.js
 */

'use strict';

// Per-call costs matching policy.yaml
const TOOL_COSTS = {
  web_search: 0.05,
  query_database: 0.1,
  call_external_api: 0.2,
  generate_report: 0.25,
};

// Session budget ceiling matching policy.yaml budget.max_cost
const BUDGET_LIMIT = 0.5;

// ---------------------------------------------------------------------------
// Minimal Agent Assembly SDK stubs used in this offline example.
// In a real integration replace these with:
//   const { AssemblyClient, BudgetPolicy } = require('@agent-assembly/sdk');
// ---------------------------------------------------------------------------

class BudgetExceededError extends Error {
  constructor(message) {
    super(message);
    this.name = 'BudgetExceededError';
  }
}

class BudgetTracker {
  constructor(maxCost) {
    this.maxCost = maxCost;
    this._spent = 0;
  }

  get spent() {
    return this._spent;
  }

  get remaining() {
    return Math.max(0, this.maxCost - this._spent);
  }

  charge(cost) {
    this._spent += cost;
  }

  status() {
    const pct = Math.round((this._spent / this.maxCost) * 100);
    return `spent=$${this._spent.toFixed(2)} / limit=$${this.maxCost.toFixed(2)} (${pct}%)`;
  }
}

class AssemblyClient {
  constructor(agentId, budget) {
    this.agentId = agentId;
    this._budget = budget;
  }

  callTool(tool, inputs = {}) {
    const cost = TOOL_COSTS[tool] ?? 0.01;

    if (cost > this._budget.remaining) {
      throw new BudgetExceededError(
        `Budget exceeded: tool '${tool}' costs $${cost.toFixed(2)} but ` +
        `only $${this._budget.remaining.toFixed(2)} remains ` +
        `(spent=$${this._budget.spent.toFixed(2)} / limit=$${this._budget.maxCost.toFixed(2)})`,
      );
    }

    this._budget.charge(cost);
    console.log(
      `  [BUDGET] charged $${cost.toFixed(2)} for '${tool}'  →  ${this._budget.status()}`,
    );
    return { status: 'ok', tool, data: `<result of ${tool}>` };
  }
}

// ---------------------------------------------------------------------------
// Example agent
// ---------------------------------------------------------------------------

function run() {
  console.log('=== Agent Assembly — Budget Limits Example ===\n');
  console.log(`Policy: max_cost=$${BUDGET_LIMIT.toFixed(2)} per session (see policy.yaml)\n`);

  const budget = new BudgetTracker(BUDGET_LIMIT);
  const client = new AssemblyClient('example-agent-001', budget);

  const calls = [
    { tool: 'web_search',        inputs: { query: 'latest AI news' } },       // $0.05 → $0.05
    { tool: 'query_database',    inputs: { table: 'customers' } },             // $0.10 → $0.15
    { tool: 'call_external_api', inputs: { endpoint: '/v1/report' } },        // $0.20 → $0.35
    { tool: 'web_search',        inputs: { query: 'weather forecast' } },     // $0.05 → $0.40
    { tool: 'generate_report',   inputs: { format: 'pdf' } },                 // $0.25 → exceeds $0.50
    { tool: 'call_external_api', inputs: { endpoint: '/v2/sync' } },          // $0.20 → blocked
  ];

  console.log('--- Running tool calls against budget ---\n');

  for (const { tool, inputs } of calls) {
    const cost = TOOL_COSTS[tool] ?? 0.01;
    const argsStr = Object.entries(inputs)
      .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
      .join(', ');
    console.log(`  → ${tool}(${argsStr})  [cost=$${cost.toFixed(2)}]`);
    try {
      const result = client.callTool(tool, inputs);
      console.log(`    ✓ allowed  →  ${JSON.stringify(result)}\n`);
    } catch (err) {
      console.log(`    ✗ denied   →  ${err.message}\n`);
    }
  }

  console.log(`\nFinal budget state: ${budget.status()}`);
}

run();
